app/src/main/java/com/duckduckgo/app/privacy/ui/WhitelistViewModel.kt

```diff
@@ -0,0 +1,120 @@
+import com.duckduckgo.app.global.SingleLiveEvent
+import com.duckduckgo.app.global.UriString
+import com.duckduckgo.app.privacy.db.UserWhitelistDao
+import com.duckduckgo.app.privacy.model.UserWhitelistedDomain
+import com.duckduckgo.app.privacy.ui.WhitelistViewModel.Command.*
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+class WhitelistViewModel(
+    private val dao: UserWhitelistDao,
+    private val dispatchers: DispatcherProvider = DefaultDispatcherProvider()
+) : ViewModel() {
+
+    data class ViewState(
+        val showWhitelist: Boolean = true,
+        val whitelist: List<UserWhitelistedDomain> = emptyList()
+    )
+
+    sealed class Command {
+        object ShowAdd : Command()
+        class ShowEdit(val entry: UserWhitelistedDomain) : Command()
+        class ConfirmDelete(val entry: UserWhitelistedDomain) : Command()
+        object ShowWhitelistFormatError : Command()
+    }
+
+    val viewState: MutableLiveData<ViewState> = MutableLiveData()
+    val command: SingleLiveEvent<Command> = SingleLiveEvent()
+
+    private val entries: LiveData<List<UserWhitelistedDomain>> = dao.all()
+    private val observer = Observer<List<UserWhitelistedDomain>> { onUserWhitelistChanged(it!!) }
+
+    init {
+        viewState.value = ViewState()
+        entries.observeForever(observer)
+    }
+
+    override fun onCleared() {
+        super.onCleared()
+        entries.removeObserver(observer)
+    }
+
+    private fun onUserWhitelistChanged(entries: List<UserWhitelistedDomain>) {
+        viewState.value = viewState.value?.copy(
+            showWhitelist = entries.isNotEmpty(),
+            whitelist = entries
+        )
+    }
+
+    fun onAddRequested() {
+        command.value = ShowAdd
+    }
+
+    fun onEntryAdded(entry: UserWhitelistedDomain) {
+        if (!UriString.isValidDomain(entry.domain)) {
+            command.value = ShowWhitelistFormatError
+            return
+        }
+        GlobalScope.launch(dispatchers.io()) {
+            addEntryToDatabase(entry)
+        }
+    }
+
+    fun onEditRequested(entry: UserWhitelistedDomain) {
+        command.value = ShowEdit(entry)
+    }
+
+    fun onEntryEdited(old: UserWhitelistedDomain, new: UserWhitelistedDomain) {
+        if (!UriString.isValidDomain(new.domain)) {
+            command.value = ShowWhitelistFormatError
+            return
+        }
+        GlobalScope.launch(dispatchers.io()) {
+            deleteEntryFromDatabase(old)
+            addEntryToDatabase(new)
+        }
+    }
+
+    fun onDeleteRequested(entry: UserWhitelistedDomain) {
+        command.value = ConfirmDelete(entry)
+    }
+
+    fun onEntryDeleted(entry: UserWhitelistedDomain) {
+        GlobalScope.launch(dispatchers.io()) {
+            deleteEntryFromDatabase(entry)
+        }
+    }
+
+    private suspend fun addEntryToDatabase(entry: UserWhitelistedDomain) {
+        withContext(dispatchers.io()) { dao.insert(entry) }
```

Comment:
We don't really need to repeat the io dispatcher here as the launch already specifies an io dispatcher - however this method was added based on PR feedback around future safety and as it is possible that a future launch could be on the main thread, I have been explicit.

Reply:
`withContext` internally checks if we are already in the same dispatcher, so there's no problem repeating here if that gives us more confidence, which I also agree is a good idea. Nice comment.